applications/staex-iod.md

```diff
@@ -0,0 +1,326 @@
+
+#### IoT devices indexing
+
+To give users possibility to find particular type of data or particular IoT device we need to index them somehow. Indexing part we want to implement through DIDs and Subsquid project.
+
+So we will implement ink! smart contract which will be deployed on Kusama network. This smart contract will have several methods:
+
+- Create DID
+- Add attribute
+- Get attribute
+- Update attribute
+- Delete attribute
+
+All these methods will produce smart contract events.
+
+To index and retrieve IoT devices (DID) we want to use Subsquid project and its API as it index all transactions and events from blockchain network and we can query them using GraphQL API from our UI. User can use public indexer or start its own. Subsquid already has an integration with Kusama.
+
+To make process of setup or update DID for IoT device easier, we will implement such logic in our provisioner. So user will need just to set some fields in config file and provisioner will sync DID attributes with network through smart contracts calls.
+
+Attributes which we want to store in DIDs:
+
+- Type of data - type of data which IoT device produces, it can be temperature, humidity or something like CCTV camera output, etc.
+- Location - it can be coordinates or name of the place.
+- Public key - device public key. We will need it later to create shared secret or know exact files location in IPFS.
+- Price to access data per KB.
+- Price to pin data per hour.
+- Device can include some other attributes but it is optional.
+
+After that we will implement a UI where users can see devices and find necessary one through filters. UI will be quite simple: header/footer + a table with device list, each row will have an information from DID. Plus we need an input or select field for filters.
+
+#### ACL
+
+Access control layer is here to restrict data usage from users which didn't pay for a data.
+
+And here is a flow: 
+
+1. User puts some tokens to smart contract for exact owner and point which type of data it needs.
+2. Provisioner on IoT device sees an event from smart contract that there is new data access request.
+3. If provisioner can produce such data.
+4. Provisioner gets tokens from smart contract.
+5. Provisioner starts to distribute data for this user.
+
+##### Data access smart contract
+
+To implement flow below we will implement a kind of multisig ink! smart contract with three methods:
+
+- Request - user requests for particular data access and sends required amount of tokens.
+- Accept - device accepts request and gets required amount of tokens.
+- Reject - device rejects request and tokens will be returned to user.
+
+Device can reject request because of:
+
+- Requested data is not available at the moment.
+- Device has high amount of request (highload).
+
+This contract will use DID smart contract to check that user sent required amount of tokens otherwise reject automatically.
```

Comment:
Would you be using an existing solution here, and if so, which one? Or would you be creating your own custom DID contract?

Reply:
We assume that it will be easier and useful to create our own smart contract because it will not so difficult and we don't need much functionality atm. Also we can make setup and integration with data access smart contract and Subsqiud more precisely.